### ISSUE TITLE
EncMode 3 and 4 BDrate regression

### ISSUE THREAD

**Issue by EwoutH** [OpenVisualCloud/SVT-AV1#734]
Between 1f82662 and 0175b58 I noticed a BDrate regressiom in EncMode 4 and 3. EncMode 5 to 8 weren't impacted as far as I can see, I didn't test mode 2 to 0 (due to compute time limitations) so they could also be impacted.

See these runs on AWCY: [EncMode 3](https://beta.arewecompressedyet.com/?job=SVT-AV1-enc-mode-3%402019-10-06&job=SVT-AV1-enc-mode-3%402019-10-22) and [EncMode 4](https://beta.arewecompressedyet.com/?job=SVT-AV1-enc-mode-4%402019-10-06&job=SVT-AV1-enc-mode-4%402019-10-22).

Here the BDrate change between aad466d and 0175b58
![Screenshot_470](https://user-images.githubusercontent.com/15776622/67384492-8875df00-f591-11e9-99df-94dfba6cc726.png)

**Comment by EwoutH** [comment]
Another serious regression in EncMode3 between 0175b58 and 7844f0b: https://beta.arewecompressedyet.com/?job=SVT-AV1-enc-mode-3%402019-10-22&job=SVT-AV1-enc-mode-3%402019-10-25

**Comment by EwoutH** [comment]
@hassount Could you give a status update on this issue?

**Comment by EwoutH** [comment]
@hguermaz @lijing0010 Could you detail in what portion #768 fixes this regression?

**Comment by EwoutH** [comment]
Since 2019-10-06 EncMode 3 [regressed](https://beta.arewecompressedyet.com/?job=SVT-AV1-enc-mode-3%402019-10-06&job=SVT-AV1-enc-mode-3%402019-11-19) only more. The [largest regression](https://beta.arewecompressedyet.com/?job=SVT-AV1-enc-mode-3%402019-10-22&job=SVT-AV1-enc-mode-3%402019-10-25) seems between 0175b58 and 7844f0b, with BDR increases between 15% and 20%. Between d8e468f and 3ea18c2 EncMode 3 performance also [regressed](https://beta.arewecompressedyet.com/?job=SVT-AV1-enc-mode-3%402019-11-05&job=SVT-AV1-enc-mode-3%402019-11-19) 2-3% depending on metrics.

EncMode 4 saw smaller but similar [regressions](https://beta.arewecompressedyet.com/?job=SVT-AV1-enc-mode-4%402019-10-06&job=SVT-AV1-enc-mode-4%402019-10-08&job=SVT-AV1-enc-mode-4%402019-11-05&job=SVT-AV1-enc-mode-4%402019-11-19) between the same commits.

@hassount @kylophone I strongly urge to assign the highest priority to this issue. EncMode 3 and 4 did offer an excellent balance between quality and speed, but are a lot less attractive to unusable with those regressions huge.

I would also suggest adding weekly runs of all EncModes to https://videocodectracker.dev/ to track potential regressions more consistently and frequently.

**Comment by EwoutH** [comment]
Any news?

**Comment by joelsole** [comment]
WIP.
Also, starting in the next ruh, SSIM and MS-SSIM will be reported in the raw data.

**Comment by hassount** [comment]
@EwoutH 

Thank you very much for this report and I apologize for the very late response on this. This issue is just not easy to solve:

We spent quite a bit of time looking into this issue and the major problem that we found is that most streams in the intervals you have pointed to along with the revisions after had corruptions and R2Rs, very few were actually decodable until #768 came in which does cause a quality degradation such is expected as the encoding speed would also increase. Now the other tricky part is that even when back porting fix to the M3-M4, we found even more R2Rs (causing up to 0.5-0.8%% BDR deviation in some clips) and therefore we needed to wait until the R2R bug fixes are in #763 #774 and #804. 

Now in the meantime, we have been working on the preset tuning in the branches: rtime-m0-test and release27_10_2019 where we are trying to get closer presets to -enc-mode 0 so -enc-mode 3/4 kept changing within our internal testing made those presets quite different from what we have in the master. The closest -enc-mode 3 that we believe would have addressed this issues and produced a closed trade-off range is the -enc-mode 3 in ms-dec1 branch.

That said, every time we try to PR the preset changes our performance and stability tests fail as -enc-mode 3/4 are still not as stable. We have tracked the root cause of the issues to the partitioning algorithm and the PRs #788 changed the partitioning decision tree drastically so we waited until that came in. That algorithm was the initial step for #858, which is a multi-staged decision of the partitioning tree which has overwritten #788 and changed the infrastructure quite a bit. As a result, we now have a relatively stable code base, and a brand new partitioning algorithm overwriting what we had before and the team is now working on rebuilding the presets (at least -enc-mode 0-3) to make sure that we have accounted for everything that was causing the instabilities in the presets. 

Long story short, because the priority was for the past few months / weeks to finish the implementation of the av1 features, and adding the new decision algorithms, the focus was on the -enc-mode 0 where all features are on and can interact with each others. Now that 95% of the features are in, we will focus on cleaning up the code and rebuild the presets. Although this is not an ideal software engineering situation, but due to all of the above changes, this issue is way too expensive for us to continue tracking as is and thus the approach above explained.

**Comment by hassount** [comment]
@EwoutH 
can we close this ?

**Comment by hassount** [comment]
Closing, let's revisit after presets are updated
